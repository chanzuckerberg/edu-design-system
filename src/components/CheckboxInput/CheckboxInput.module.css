/* ------------------------------------*\
    #CHECKBOX INPUT
\*------------------------------------ */

/**
 * Wrapper for the checkbox, including the visually hidden input element.
 * 1) Setting position is required to place the absolutely positioned input element in the same spot as the SVG 
 *    when a (n^c)parent element in the tree is relatively positioned.
 * 2) line-height is set to be same as checkbox size to equal checkbox size and prevent browser styling.
 */
.input__wrapper {
  position: relative; /* 1 */
  line-height: var(--checkbox-svg-size); /* 2 */
}

/**
   * The visually hidden input element for the checkbox. The visual checkmark is provided by an svg element.
   * 1) Removes default margins placed by browser for checkboxes.
   * 2) Remove the checkbox from the page flow, positioning it on top of the SVG.
   * 3) Set same dimensions as the CheckboxSvg element.
   * 4) Hide the input element visually.
   *    This ensures the checkbox is still "physically" present so that all users,
   *    especially on touch screen readers, still interact with the real checkbox element
   *    where it would naturally be present.
   */
.checkbox__input {
  cursor: pointer;
  margin: 0; /* 1 */
  position: absolute; /* 2 */
  height: var(--checkbox-svg-size); /* 3 */
  width: var(--checkbox-svg-size);
  opacity: 0; /* 4 */
}

/**
   * The visible checkbox graphic.
   * 1) Provides a focus indicator outline around the checkbox.
   */
.checkbox__input + svg {
  vertical-align: top;
  display: inline-block;
  color: var(--eds-color-neutral-500);
  --checked-color: var(--eds-color-brand-600);
}

.checkbox__input:checked + svg > .checkmark__outline {
  fill: var(--checked-color);
  stroke: var(--checked-color);
}

.checkbox__input:checked + svg > .checkmark__content {
  fill: var(--eds-color-white);
}

.checkbox__input:focus + svg {
  @apply outline-none rounded ring-2 ring-info-400; /* 1 */
}

/**
   * Disabled styles 
   */
.checkbox__input:disabled {
  cursor: not-allowed;
}

.checkbox__input:disabled + svg {
  color: var(--eds-color-neutral-300);
  --checked-color: var(--eds-color-neutral-300);
}
