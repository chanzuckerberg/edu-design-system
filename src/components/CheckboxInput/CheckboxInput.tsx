import clsx from 'clsx';
import React from 'react';
import useForwardedRef from '../../util/useForwardedRef';
import styles from './CheckboxInput.module.css';

type CheckboxHTMLElementProps = Omit<
  React.InputHTMLAttributes<HTMLInputElement>,
  'checked' | 'id' | 'size'
>;

export type CheckboxInputProps = CheckboxHTMLElementProps & {
  /**
   * Whether checkbox is checked.
   */
  checked?: boolean;
  /**
   * Additional classnames passed in for styling.
   */
  className?: string;
  /**
   * Checkbox ID. Used to connect the input with a label for accessibility purposes.
   */
  id: string;
  /**
   * Whether the checkbox is "indeterminate". Neither checked nor unchecked. The most common use
   * case for this is when a checkbox has sub-checkboxes, to represent a "partially checked" state.
   */
  indeterminate?: boolean;
};

/**
 * Checkbox input element, exported for greater flexibility.
 * You must provide an `id` prop and connect it to a visible label.
 */
export const CheckboxInput = React.forwardRef<
  HTMLInputElement,
  CheckboxInputProps
>(({ checked, className, disabled, indeterminate, ...other }, ref) => {
  const forwardedRef = useForwardedRef(ref);

  // Make this checkbox indeterminate. Can only be done with JS for some reason.
  // See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#indeterminate_state_checkboxes.
  React.useEffect(() => {
    if (forwardedRef.current) {
      forwardedRef.current.indeterminate = !!indeterminate;
    }
  }, [forwardedRef, indeterminate]);

  return (
    <input
      checked={checked}
      className={clsx(className, styles['checkbox__input'])}
      disabled={disabled}
      ref={forwardedRef}
      type="checkbox"
      {...other}
    />
  );
});

CheckboxInput.displayName = 'CheckboxInput';
