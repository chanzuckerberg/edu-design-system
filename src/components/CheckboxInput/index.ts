export { CheckboxInput as default, svgStyle } from './CheckboxInput';
export type { CheckboxInputProps } from './CheckboxInput';
