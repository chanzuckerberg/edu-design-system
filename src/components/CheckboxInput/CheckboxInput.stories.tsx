import { StoryObj, Meta } from '@storybook/react';
import React from 'react';
import { CheckboxInput } from './CheckboxInput';

export default {
  title: 'Atoms/Forms/CheckboxInput',
  component: CheckboxInput,
  argTypes: {
    // For some reason Storybook does not infer all `checked` correctly;
    checked: {
      control: 'radio',
      options: [true, false, 'indeterminate'],
    },
  },
  args: {
    'aria-label': 'checkbox',
    id: 'checkbox',
  },
  decorators: [
    (Story) => (
      <div
        style={{
          margin: '0.25rem', // Provides spacing to see focus indicator around checkbox.
        }}
      >
        <Story />
      </div>
    ),
  ],
} as Meta;

type Args = React.ComponentProps<typeof CheckboxInput>;
export const Default: StoryObj<Args> = {};

/**
 * Controlled example to make checked stories interactive.
 */
function CheckedExample(args: Args) {
  const [checked, setChecked] = React.useState(true);
  const handleChange = () => {
    setChecked(!checked);
  };

  return <CheckboxInput checked={checked} onChange={handleChange} {...args} />;
}

export const Checked: StoryObj<Args> = {
  render: () => <CheckedExample id="checked-example" />,
};

export const Indeterminate: StoryObj<Args> = {
  ...Default,
  args: {
    checked: 'indeterminate',
    readOnly: true,
  },
};

export const Disabled: StoryObj<Args> = {
  ...Default,
  render: (args) => (
    <div>
      <CheckboxInput disabled {...args} id="disabled" />
      <br />
      <CheckboxInput checked disabled {...args} id="disabled-checked" />
      <br />
      <CheckboxInput
        checked="indeterminate"
        disabled
        {...args}
        id="disabled-indeterminate"
      />
    </div>
  ),
};
