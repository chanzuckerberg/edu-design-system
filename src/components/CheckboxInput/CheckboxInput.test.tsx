import { generateSnapshots } from '@chanzuckerberg/story-utils';
import * as stories from './CheckboxInput.stories';

describe('<CheckboxInput />', () => {
  generateSnapshots(stories);
});
