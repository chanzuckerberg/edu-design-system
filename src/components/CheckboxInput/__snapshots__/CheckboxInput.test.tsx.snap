// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxInput /> Checked story renders snapshot 1`] = `
<div
  style="margin: 0.25rem;"
>
  <span
    class="input__wrapper"
  >
    <input
      checked=""
      class="checkbox__input"
      data-bootstrap-override="checkbox"
      id="checked-example"
      style="--checkbox-svg-size: 1.25rem;"
      type="checkbox"
    />
    <svg
      aria-hidden="true"
      height="1.25rem"
      viewBox="0 0 20 20"
      width="1.25rem"
    >
      <rect
        class="checkmark__outline"
        fill="none"
        height="18"
        rx="3"
        stroke="currentColor"
        stroke-width="2"
        width="18"
        x="1"
        y="1"
      />
      <path
        class="checkmark__content"
        d="M15.5605 5.2094C15.447 5.12571 15.3181 5.06522 15.1811 5.03137C15.0442 4.99752 14.902 4.99099 14.7625 5.01214C14.6231 5.0333 14.4892 5.08172 14.3684 5.15465C14.2477 5.22758 14.1425 5.32359 14.0589 5.43718L8.96703 12.3478L5.72917 9.7572C5.50592 9.58618 5.22455 9.50943 4.94539 9.54341C4.66622 9.57738 4.41148 9.71937 4.23576 9.93893C4.06004 10.1585 3.97733 10.4382 4.00536 10.718C4.03339 10.9978 4.16993 11.2555 4.38572 11.4359L8.49687 14.7244C8.61048 14.8133 8.74065 14.8786 8.87979 14.9166C9.01893 14.9546 9.16424 14.9645 9.30724 14.9457C9.45024 14.9269 9.58805 14.8798 9.71264 14.8071C9.83723 14.7344 9.94609 14.6377 10.0329 14.5225L15.7917 6.71187C15.8753 6.59806 15.9356 6.4689 15.9691 6.33178C16.0027 6.19465 16.009 6.05225 15.9875 5.91272C15.966 5.77318 15.9173 5.63925 15.844 5.51857C15.7707 5.39789 15.6744 5.29283 15.5605 5.2094Z"
        fill="none"
      />
    </svg>
  </span>
</div>
`;

exports[`<CheckboxInput /> Default story renders snapshot 1`] = `
<div
  style="margin: 0.25rem;"
>
  <span
    class="input__wrapper"
  >
    <input
      aria-label="checkbox"
      class="checkbox__input"
      data-bootstrap-override="checkbox"
      id="checkbox"
      style="--checkbox-svg-size: 1.25rem;"
      type="checkbox"
    />
    <svg
      aria-hidden="true"
      height="1.25rem"
      viewBox="0 0 20 20"
      width="1.25rem"
    >
      <rect
        class="checkmark__outline"
        fill="none"
        height="18"
        rx="3"
        stroke="currentColor"
        stroke-width="2"
        width="18"
        x="1"
        y="1"
      />
      <path
        class="checkmark__content"
        d="M15.5605 5.2094C15.447 5.12571 15.3181 5.06522 15.1811 5.03137C15.0442 4.99752 14.902 4.99099 14.7625 5.01214C14.6231 5.0333 14.4892 5.08172 14.3684 5.15465C14.2477 5.22758 14.1425 5.32359 14.0589 5.43718L8.96703 12.3478L5.72917 9.7572C5.50592 9.58618 5.22455 9.50943 4.94539 9.54341C4.66622 9.57738 4.41148 9.71937 4.23576 9.93893C4.06004 10.1585 3.97733 10.4382 4.00536 10.718C4.03339 10.9978 4.16993 11.2555 4.38572 11.4359L8.49687 14.7244C8.61048 14.8133 8.74065 14.8786 8.87979 14.9166C9.01893 14.9546 9.16424 14.9645 9.30724 14.9457C9.45024 14.9269 9.58805 14.8798 9.71264 14.8071C9.83723 14.7344 9.94609 14.6377 10.0329 14.5225L15.7917 6.71187C15.8753 6.59806 15.9356 6.4689 15.9691 6.33178C16.0027 6.19465 16.009 6.05225 15.9875 5.91272C15.966 5.77318 15.9173 5.63925 15.844 5.51857C15.7707 5.39789 15.6744 5.29283 15.5605 5.2094Z"
        fill="none"
      />
    </svg>
  </span>
</div>
`;

exports[`<CheckboxInput /> Disabled story renders snapshot 1`] = `
<div
  style="margin: 0.25rem;"
>
  <div>
    <span
      class="input__wrapper input__wrapper--disabled"
    >
      <input
        aria-label="checkbox"
        class="checkbox__input"
        data-bootstrap-override="checkbox"
        disabled=""
        id="disabled"
        style="--checkbox-svg-size: 1.25rem;"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        height="1.25rem"
        viewBox="0 0 20 20"
        width="1.25rem"
      >
        <rect
          class="checkmark__outline"
          fill="none"
          height="18"
          rx="3"
          stroke="currentColor"
          stroke-width="2"
          width="18"
          x="1"
          y="1"
        />
        <path
          class="checkmark__content"
          d="M15.5605 5.2094C15.447 5.12571 15.3181 5.06522 15.1811 5.03137C15.0442 4.99752 14.902 4.99099 14.7625 5.01214C14.6231 5.0333 14.4892 5.08172 14.3684 5.15465C14.2477 5.22758 14.1425 5.32359 14.0589 5.43718L8.96703 12.3478L5.72917 9.7572C5.50592 9.58618 5.22455 9.50943 4.94539 9.54341C4.66622 9.57738 4.41148 9.71937 4.23576 9.93893C4.06004 10.1585 3.97733 10.4382 4.00536 10.718C4.03339 10.9978 4.16993 11.2555 4.38572 11.4359L8.49687 14.7244C8.61048 14.8133 8.74065 14.8786 8.87979 14.9166C9.01893 14.9546 9.16424 14.9645 9.30724 14.9457C9.45024 14.9269 9.58805 14.8798 9.71264 14.8071C9.83723 14.7344 9.94609 14.6377 10.0329 14.5225L15.7917 6.71187C15.8753 6.59806 15.9356 6.4689 15.9691 6.33178C16.0027 6.19465 16.009 6.05225 15.9875 5.91272C15.966 5.77318 15.9173 5.63925 15.844 5.51857C15.7707 5.39789 15.6744 5.29283 15.5605 5.2094Z"
          fill="none"
        />
      </svg>
    </span>
    <br />
    <span
      class="input__wrapper input__wrapper--disabled"
    >
      <input
        aria-label="checkbox"
        checked=""
        class="checkbox__input"
        data-bootstrap-override="checkbox"
        disabled=""
        id="disabled-checked"
        style="--checkbox-svg-size: 1.25rem;"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        height="1.25rem"
        viewBox="0 0 20 20"
        width="1.25rem"
      >
        <rect
          class="checkmark__outline"
          fill="none"
          height="18"
          rx="3"
          stroke="currentColor"
          stroke-width="2"
          width="18"
          x="1"
          y="1"
        />
        <path
          class="checkmark__content"
          d="M15.5605 5.2094C15.447 5.12571 15.3181 5.06522 15.1811 5.03137C15.0442 4.99752 14.902 4.99099 14.7625 5.01214C14.6231 5.0333 14.4892 5.08172 14.3684 5.15465C14.2477 5.22758 14.1425 5.32359 14.0589 5.43718L8.96703 12.3478L5.72917 9.7572C5.50592 9.58618 5.22455 9.50943 4.94539 9.54341C4.66622 9.57738 4.41148 9.71937 4.23576 9.93893C4.06004 10.1585 3.97733 10.4382 4.00536 10.718C4.03339 10.9978 4.16993 11.2555 4.38572 11.4359L8.49687 14.7244C8.61048 14.8133 8.74065 14.8786 8.87979 14.9166C9.01893 14.9546 9.16424 14.9645 9.30724 14.9457C9.45024 14.9269 9.58805 14.8798 9.71264 14.8071C9.83723 14.7344 9.94609 14.6377 10.0329 14.5225L15.7917 6.71187C15.8753 6.59806 15.9356 6.4689 15.9691 6.33178C16.0027 6.19465 16.009 6.05225 15.9875 5.91272C15.966 5.77318 15.9173 5.63925 15.844 5.51857C15.7707 5.39789 15.6744 5.29283 15.5605 5.2094Z"
          fill="none"
        />
      </svg>
    </span>
    <br />
    <span
      class="input__wrapper input__wrapper--disabled"
    >
      <input
        aria-checked="mixed"
        aria-label="checkbox"
        checked=""
        class="checkbox__input"
        data-bootstrap-override="checkbox"
        disabled=""
        id="disabled-indeterminate"
        style="--checkbox-svg-size: 1.25rem;"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        height="1.25rem"
        viewBox="0 0 20 20"
        width="1.25rem"
      >
        <rect
          class="checkmark__outline"
          fill="none"
          height="18"
          rx="3"
          stroke="currentColor"
          stroke-width="2"
          width="18"
          x="1"
          y="1"
        />
        <path
          class="checkmark__content"
          d="M 14.00,9.00 C 14.55,9.00 15.00,9.45 15.00,10.00 15.00,10.00 15.00,10.00 15.00,10.00 15.00,10.55 14.55,11.00 14.00,11.00 14.00,11.00 6.00,11.00 6.00,11.00 5.45,11.00 5.00,10.55 5.00,10.00 5.00,10.00 5.00,10.00 5.00,10.00 5.00,9.45 5.45,9.00 6.00,9.00 6.00,9.00 14.00,9.00 14.00,9.00 Z"
          fill="none"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`<CheckboxInput /> Indeterminate story renders snapshot 1`] = `
<div
  style="margin: 0.25rem;"
>
  <span
    class="input__wrapper"
  >
    <input
      aria-checked="mixed"
      aria-label="checkbox"
      checked=""
      class="checkbox__input"
      data-bootstrap-override="checkbox"
      id="checkbox"
      readonly=""
      style="--checkbox-svg-size: 1.25rem;"
      type="checkbox"
    />
    <svg
      aria-hidden="true"
      height="1.25rem"
      viewBox="0 0 20 20"
      width="1.25rem"
    >
      <rect
        class="checkmark__outline"
        fill="none"
        height="18"
        rx="3"
        stroke="currentColor"
        stroke-width="2"
        width="18"
        x="1"
        y="1"
      />
      <path
        class="checkmark__content"
        d="M 14.00,9.00 C 14.55,9.00 15.00,9.45 15.00,10.00 15.00,10.00 15.00,10.00 15.00,10.00 15.00,10.55 14.55,11.00 14.00,11.00 14.00,11.00 6.00,11.00 6.00,11.00 5.45,11.00 5.00,10.55 5.00,10.00 5.00,10.00 5.00,10.00 5.00,10.00 5.00,9.45 5.45,9.00 6.00,9.00 6.00,9.00 14.00,9.00 14.00,9.00 Z"
        fill="none"
      />
    </svg>
  </span>
</div>
`;
